src/answers.sql

```diff
@@ -1,15 +1,250 @@
+	ACCOUNTS
+ORDER BY
+	MOUNT DESC
+LIMIT
+	5;
 
+--
+--
+--
+--
 -- 4
+-- I MAKE SURE I USE EVERY ACCOUNT BY UNION ALL!
+-- INNER JOIN MAY LEAVE OUT USERS WITH NO MOVEMENTS
+CREATE
+OR REPLACE FUNCTION GET_ACCOUNTS_FINAL_BALANCE () RETURNS TABLE (ACCOUNT_ID UUID, MONEY NUMERIC(15, 2)) AS $$ BEGIN RETURN QUERY WITH accounts_money_movements AS (
+        (
+            -- account_from transactions
+            SELECT account_from AS account_id,
+                CASE
+                    WHEN type = 'IN' THEN mount
+                    ELSE mount * -1
+                END AS mount
+            FROM movements
+        )
+        UNION ALL
+        (
+            -- account_to transactions
+            SELECT CASE
+                    WHEN type = 'TRANSFER' THEN account_to
+                    ELSE account_from
+                END AS account_id,
+                CASE
+                    WHEN type = 'TRANSFER' THEN mount
+                    ELSE 0
+                END AS mount
+            FROM movements
+        )
+        UNION ALL
+        (
+            -- accounts starting money
+            SELECT id AS account_id,
+                mount
+            FROM accounts
+        )
+    )
+SELECT am.account_id,
+    ROUND(SUM(am.mount)::NUMERIC, 2) AS money
+FROM accounts_money_movements am
+GROUP BY am.account_id;
+END;
+$$ LANGUAGE PLPGSQL;
 
+--
+-- Solution applying function  get_accounts_final_balance
+--
+SELECT
+	U.ID,
+	U.NAME,
+	SUM(AB.MONEY) ALL_MONEY
+FROM
+	GET_ACCOUNTS_FINAL_BALANCE () AB
+	INNER JOIN ACCOUNTS A ON AB.ACCOUNT_ID = A.ID
+	INNER JOIN USERS U ON U.ID = A.USER_ID
+GROUP BY
+	U.ID,
+	U.NAME
+ORDER BY ALL_MONEY DESC
+LIMIT 3;
+
+--
+--
+--
 -- 5
+--
+DO $$
+DECLARE
+    account_balance_a DOUBLE PRECISION;
+    account_balance_b DOUBLE PRECISION;
+	
+    account_id_a UUID := '3b79e403-c788-495a-a8ca-86ad7643afaf';
+    account_id_b UUID := 'fd244313-36e5-4a17-a27c-f8265bc46590';
+BEGIN
+
+	--a. First, get the ammount for the account `3b79e403-c788-495a-a8ca-86ad7643afaf` and `fd244313-36e5-4a17-a27c-f8265bc46590` after all their movements.
+	
+	SELECT money into account_balance_a
+	FROM get_accounts_final_balance() 
+	WHERE account_id = account_id_a;
+
+	SELECT money into account_balance_b
+	FROM get_accounts_final_balance() 
+	WHERE account_id = account_id_b;
+
+	RAISE NOTICE 'Account balance from % is %', account_id_a, account_balance_a;
+	RAISE NOTICE 'Account balance from % is %', account_id_b, account_balance_b;
+
+	--b. Add a new movement with the information: from: 3b79e403-c788-495a-a8ca-86ad7643afaf make a transfer to fd244313-36e5-4a17-a27c-f8265bc46590 mount: 50.75
+	IF account_balance_a < 50.75 THEN
+		RAISE EXCEPTION 'Account balance is insufficient!';
+	ELSE 
+		INSERT INTO movements  (id, type, account_from, account_to, mount) 
+		VALUES (gen_random_uuid(), 'TRANSFER', account_id_a,account_id_b, 50.75);
+		RAISE NOTICE 'Transfer % USD from % to % MADE SUCCESSFULLY!', 50.75, account_id_a, account_id_b;
+	END IF;
+
```

Comment:
You could have add a `savepoint` here, to save your changes up to this point

Reply:
I could, but it'd loose the point of the transaction!

In case you may want to have a save point, I would most likely make 2 transactions.
This way it commits itself once the first is done or the second is finished.


TRANSACTION 1 for TRANSFER movement:

```sql

DO $$
DECLARE
    account_balance_a DOUBLE PRECISION;
	
    account_id_a UUID := '3b79e403-c788-495a-a8ca-86ad7643afaf';
    account_id_b UUID := 'fd244313-36e5-4a17-a27c-f8265bc46590';

    transfer_mount DOUBLE PRECISION := 50.75;
BEGIN

EXCEPTION
--
END;
$$;

```


TRANSACTION 2 for OUT movement:

```sql

DO $$
DECLARE
    account_balance_a DOUBLE PRECISION;
	
    account_id_a UUID := '3b79e403-c788-495a-a8ca-86ad7643afaf';
    account_id_b UUID := 'fd244313-36e5-4a17-a27c-f8265bc46590';

    out_mount DOUBLE PRECISION := 50.75;
BEGIN

EXCEPTION
--
END;
$$;

```